### DISCUSSION TITLE
Datasets and git submodules

### DISCUSSION THREAD

**Discussion by ampx-mg** [apache/airflow#38261] [Q&A]
Hello,

we have deployed airflow using the official helm chart on EKS, we use `KubernetesExecutor` with `git-sync` side containers. Our project setup is one git repo per project + 1 core repository with all the project repos added as git submodules.  
We would like to start out with data-aware scheduling and got stuck pretty fast. We have some jobs which download data for multiple projects, so these jobs sit in a special repository. Now my question is, where should I specify the `Dataset` definition? Producer of the data is in different git repository than the consumer of said data.  
I figured I could possibly copy paste the `Dataset` across the projects, where needed, but maybe there is a more elegant solution.

**Comment by ampx-mg** [comment]
I was thinking about this more last night and could come up with a couple of solutions, but each of them has a drawback:

- module inside the core repo - now I would have to include the core repo in every project CI/CD pipeline, because we test each project independently
- module inside each project's repo - a lot of copy pasting, not convinced this is the best idea
- separate python package - not sure if this would work, I am concerned whether airflow can pick up the Datasets seamlessly. Also how to handle cases if one project has different version of the package than others?
- git submodule added to each repo - again the concern about different checked out commits is present

Any feedback is appreciated!

**Reply by potiuk** [reply]
It's the usual question and the usual answer - depends what you want to achieve - if you have common place, you get more DRY but introduce more coupling, if you copy & paste - you have less coupling and less DRY. 

If you have a lot of datasets and are afraid of typos, common module imported by others (similar to structure described in https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/modules_management.html ) is a good idea. 

BTW. You migh want to comment and review https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-67+Multi-tenant+deployment+of+Airflow+components proposal as it seems to fit your case nicely. And there is also a question about dataset access rules that you might be a good user to comment on.

**Reply by potiuk** [reply]
(so in short - yes there are always trade-off and drawbacks, you just have to pick the drawbacks you can live with)
